src/main/java/org/opensearch/neuralsearch/processor/NormalizationProcessorWorkflow.java

```diff
@@ -107,16 +136,43 @@ private void updateOriginalQueryResults(final List<QuerySearchResult> querySearc
                 )
             );
         }
-        for (int index = 0; index < querySearchResults.size(); index++) {
-            QuerySearchResult querySearchResult = querySearchResults.get(index);
-            CompoundTopDocs updatedTopDocs = queryTopDocs.get(index);
-            float maxScore = updatedTopDocs.getTotalHits().value > 0 ? updatedTopDocs.getScoreDocs().get(0).score : 0.0f;
+        return queryTopDocs;
+    }
 
-            // create final version of top docs with all updated values
-            TopDocs topDocs = new TopDocs(updatedTopDocs.getTotalHits(), updatedTopDocs.getScoreDocs().toArray(new ScoreDoc[0]));
+    /**
+     * Get Max score on Shard
+     * @param updatedTopDocs updatedTopDocs compound top docs on a shard
+     * @param isSortEnabled if sort is enabled or disabled
+     * @return  max score
+     */
+    private float maxScoreForShard(CompoundTopDocs updatedTopDocs, boolean isSortEnabled) {
+        if (updatedTopDocs.getTotalHits().value == 0 || updatedTopDocs.getScoreDocs().isEmpty()) {
+            return MAX_SCORE_WHEN_NO_HITS_FOUND;
+        }
+        if (isSortEnabled) {
+            float maxScore = MAX_SCORE_WHEN_NO_HITS_FOUND;
+            // In case of sorting iterate over score docs and deduce the max score
+            for (ScoreDoc scoreDoc : updatedTopDocs.getScoreDocs()) {
+                maxScore = Math.max(maxScore, scoreDoc.score);
+            }
+            return maxScore;
+        } else {
```

Comment:
nit: You can remove the else

Reply:
I would say lets keep this else because it gives more readability for someone who is new and trying to understand the code. It helps to explain clearly what happens when sorting is not applied. Removing else, would look aesthetically better, code readability would decrease. Considering the complexity of this whole feature, I would suggest to give priority to code readability.